src/java/com/puppetlabs/ssl_utils/SSLUtils.java

```diff
@@ -438,6 +445,60 @@ public static X509CRL revokeMultiple(X509CRL crl,
         return buildCRL(crl, issuerPrivateKey, issuerPublicKey, builder);
     }
 
+    /**
+     * Given a certificate chain and CRLs, validate CRLS and all certs in the chain.
+     *
+     * @param certs The certificate chain to validate
+     * @param crls The CRL chain to validate
+     * @throws CertificateException
+     * @throws CertPathValidatorException
+     * @throws IOException
+     * @throws InvalidAlgorithmParameterException
+     * @throws KeyStoreException
+     * @throws NoSuchAlgorithmException
+     * @throws NoSuchProviderException
+     */
+    public static void validateCertChain(List<X509Certificate> certs,
+                                       List<X509CRL> crls)
+        throws CertificateException, CertPathValidatorException, IOException,
+               InvalidAlgorithmParameterException, KeyStoreException,
+               NoSuchAlgorithmException, NoSuchProviderException {
+        // there are bouncycastle versions of CertificateFactory and CertPath
+        // that I haven't looked into - not sure when using bouncycastle vs. java
+        // makes sense
+        final CertificateFactory certFactory = CertificateFactory.getInstance("X.509");
+        final CertPath certPath = certFactory.generateCertPath(certs);
+        final CertPathValidator validator;
+        if (isFIPS()) {
+            validator = CertPathValidator.getInstance(PKIX_KEYMANAGER_ALGO, BOUNCYCASTLE_JSSE_PROVIDER);
+        } else {
+            validator = CertPathValidator.getInstance(CertPathValidator.getDefaultType());
+        }
+        final KeyStore truststore = certsToTrustStore(certs);
+        PKIXRevocationChecker revocationChecker = (PKIXRevocationChecker)validator.getRevocationChecker();
+        // we might want additional options, like SOFT_FAIL, but not sure
+        // see https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/cert/PKIXRevocationChecker.Option.html
```

Comment:
I _think_ we won't need that, because we're not trying to fetch the CRL over the network (we don't use CRL distribution points).

Reply:
The other two options you're using here look right.

Reply:
sweet